### ISSUE TITLE
Assembly polishing

### ISSUE THREAD

**Issue by jasperlinthorst** [PacificBiosciences/GenomicConsensus#72]
Hi all,
I have a pacbio assembly of a large genome. I have used pbalign (with the --forQuiver option) to align the raw reads back to the assembly. This resulted in many cmp.h5 files which should contain the information needed for quiver to generate a polished consensus of my assembly. What is at this point the best way to proceed?

When I try to run quiver by specifying a fofn with all cmp.h5 files, it looks like quiver is trying to load all cmp.h5 files into memory, which eventually makes the process run out of memory. When I merged all cmp.h5 files and tried to run quiver on this single huge file I was running into this exception: "AttributeError: 'record' object has no attribute 'StartRow'", which is probably caused by some kind of version mismatch with pbcore.

I don't have access to a SMRT portal, so I can't use that. Is there any documentation on which combination of versions of the different tools I should install in order to polish my assembly without using the smrt portal? Or, are my efforts not going to lead anywhere because I should apply the alignment/polishing step per contig? Please advise...

Thanks,
Jasper

Ps, here is some version information on the different tools that I have installed right now:

pbalign 0.2.0.141024
quiver 2.0.0
pbcore-1.2.8
blasr 3.0.0

**Comment by txje** [comment]
Hi Jasper,
I ran into the same problem - the "StartRow" error occurs when the cmp.h5 file is not sorted, as quiver requires. You can use cmph5tools.py to sort the cmp.h5. Some combination of merging and sorting also seems to drop the chemistry information again, so you'll want to use loadChemistry.py to add the chemistry back to the merged, sorted cmp.h5 file. All that said, I still can't get quiver to complete - seems to be deadlocked in the multithreading code...
Jeremy

**Comment by dalexander** [comment]
Thanks for helping out, @txje.  pbalign is supposed to do all the sorting and chemistry loading for you.  If you paste your pbalign command maybe we could see what went wrong.

@txje, can you report your deadlock issue as a separate issue here?

**Comment by jasperlinthorst** [comment]
@txje Thanks Jeremy, that indeed fixed the error... Thanks!
@dalexander I run pbalign for every smrtcell as follows:
pbalign $smrtcell.bas.h5 p_ctg.fasta $smrtcell.cmp.h5 --forQuiver --nproc 24 --tmpDir=/scratch/shared -vv

This results in multiple cmp.h5 files which I then merged using cmph5tools:
cmph5tools.py merge --outFile ALL.cmp.h5 *.cmp.h5
Thanks to Jeremy I then added a sort step:
cmph5tools.py sort --deep ALL.cmp.h5
This increases the filesize of ALL.cmp.h5 file with a factor 3.
I now run quiver as follows:
quiver --verbose -j32 ALL.cmp.h5 -r p_ctg.fasta -o variants.gff -o polished.fasta

This now seems to run fine...

[INFO] Using pbcommand v0.3.21
[INFO] completed setting up logger with <functools.partial object at 0x2af6a2c8d838>
[INFO] log opts {'file_name': None, 'level': 20}
[INFO] h5py version: 2.5.0
[INFO] hdf5 version: 1.8.10
[INFO] ConsensusCore version: 1.0.1
[INFO] ConsensusCore2 version: ConsensusCore2 unavailable
[INFO] Starting.
[INFO] Peeking at file ALL.cmp.h5
[INFO] Input data: numAlnHits=25452238
[INFO] Loading reference
[INFO] Loaded 6313 of 6313 reference groups from p_ctg.fasta 
[INFO] Will use quiver algorithm
[INFO] Using Quiver parameters file ...
[INFO] Using Quiver parameter set(s): P6-C4.AllQVsMergingByChannelModel, P4-C2.AllQVsMergingByChannelModel
[INFO] Available CPUs: 32
[INFO] Requested workers: 32
[INFO] Parallel Mode: Process
[INFO] Launched compute slaves.
[INFO] Launched collector slave.
slurmstepd: Job 2076677 exceeded memory limit (268236868 > 264316928), being killed
slurmstepd: **Exceeded job memory limit**


...apart from the fact that the process runs out of the available 256GB of RAM again...

How should I split up this task so that I can manage to run it within 256GB?

Thanks for your reply!
